lib/zoho_auth_client.rb

```diff
@@ -0,0 +1,40 @@
+class ZohoAuthClient
+  ACCESS_TOKEN_REQUEST_ENDPOINT = 'https://accounts.zoho.com/oauth/v2/token'.freeze
+
+  CLIENT_ID = ArchiveConfig.ZOHO_CLIENT_ID
+  CLIENT_SECRET = ArchiveConfig.ZOHO_CLIENT_SECRET
+  ZOHO_REFRESH_TOKEN = ArchiveConfig.ZOHO_REFRESH_TOKEN
+
+  # TODO: Can we get rid of this? We don't redirect, but it has to match
+  # whatever was filled in when creating the original grant_code
+  REDIRECT_URI = 'https://archiveofourown.org/support'.freeze
+
+  # TODO: Can we reduce this?
+  SCOPE = 'Desk.tickets.ALL,Desk.contacts.READ,Desk.contacts.WRITE,Desk.contacts.CREATE,Desk.basic.READ,Desk.search.READ'.freeze
+
+  def initialize
+    @access_token = access_token
+  end
+
+  def access_token
+    response_raw = HTTParty.post(ACCESS_TOKEN_REQUEST_ENDPOINT + '?' + query_string(access_token_params))
+    JSON.parse(response_raw.body)['access_token']
```

Comment:
The access token typically lasts an hour, according to the docs. I'm guessing there's usually more than one support request/abuse report per hour, so to reduce the overhead, maybe this should use memcached or redis to cache the token, and refresh only when necessary? Or is there a particular reason you want to refresh each time?

The response is supposed to return the number of seconds that the token will be good for, so something like this might work if you want to cache with memcached:
```
return cached_token if (cached_token = Rails.cache.read("zoho_access_token")).present?
response_raw = HTTParty.post(ACCESS_TOKEN_REQUEST_ENDPOINT + '?' + query_string(access_token_params))
response_json = JSON.parse(response_raw.body)
# Subtract extra time to make sure that the token doesn't expire at an inconvenient time:
cache_time = response_json["expires_in_sec"] - 1.minute
Rails.cache.write("zoho_access_token", response_json["access_token"], expires_in: cache_time)
```

Reply:
We should probably set [`race_condition_ttl`](https://api.rubyonrails.org/v5.1.6.2/classes/ActiveSupport/Cache/Store.html#method-i-fetch) on the cache block as well.

Reply:
Doesn't that only work when calling `fetch`?

(I didn't write this as a `fetch` block because `expires_in` depends on the value that would be calculated inside the cache block. If there were a cache block, that is.)

Reply:
The option works is there for `write` too, `fetch` just passes it along.

Agree on not having cache blocks though.

Reply:
But `race_condition_ttl` can only be applied properly when reading from the cache (and discovering that it's expired), so setting it with `write` doesn't actually do anything except allow it to be used within `fetch`. (Since it's not stored alongside the cache entry.)

The only place in the rails code that it's actually applied is in `handle_expired_entry`, which is only called by `fetch`.

Reply:
:+1: yeah, we don't need it. Thanks for catching that.